content/blog/2016/2016-06-29-from-freestyle-to-pipeline.adoc

```diff
@@ -0,0 +1,213 @@
+
+
+---
+
+
+When we migrated to newer infrastructure for
+link:https://ci.jenkins.io[ci.jenkins.io] earlier this year, instead of
+migrating jobs verbatim, I took the opportunity to prune stale jobs and refactor
+a number of others into proper link:/solutions/pipeline[Pipelines], statistics
+generation being an obvious target!
+
+Our requirements for statistics generation, in their most basic form, are:
+
+* Enable a sequence of dependent tasks to be executed as a logical group (a
+  pipeline)
+* Enable executing those dependent tasks on various pieces of infrastructure
+  which support different requirements
+* Actually generate those pretty blue charts
+
+
+[NOTE]
+--
+If you wish to skip ahead, you can jump straight to the
+link:https://github.com/jenkins-infra/infra-statistics/blob/a6dcaa29fca9a4f61143954fb9e1300c2f995a89/Jenkinsfile[Jenkinsfile]
+which implements our new Pipeline.
+--
+
+
+The first iteration of the `Jenkinsfile` simply defined the conceptual stages we
+would need:
+
+[source, groovy]
+----
+
+node {
+    stage 'Sync raw data and census files'
+
+    stage 'Process raw logs'
+
+    stage 'Generate census data'
+
+    stage 'Generate stats'
+
+    stage 'Publish census'
+
+    stage 'Publish stats'
+}
+----
+
+How exciting! But not terrifically useful. With the implementation of the first
+couple stages I noticed that the Pipeline might be syncing _dozens_ of gigabytes
+of data every time it ran on a new agent in the cluster. While this problem
+will soon be solved by the
+link:https://github.com/jenkinsci/external-workspace-manager-plugin[External
+Workspace Manager plugin], which is currently being developed, I needed to pin
+the execution to a consistent agent.
+
+
+[source, groovy]
+----
+/* `census` is a node label for a single machine, ideally, which will be
+ * consistently used for processing usage statistics and generating census data
+ */
+node('census && docker') {
+    /* .. */
+}
+----
+
+
+Restricting a workload which previously used multiple agents, to a single one,
+introduced the next challenge. As an infrastructure administrator, technically
+speaking, I _could_ just install all the system dependencies that I want on this
+one special Jenkins agent. But what kind of example would that be setting!
+
+This process was going to require:
+
+* JDK8
+* link:https://www.groovy-lang.org[Groovy]
+* A running link:https://www.mongodb.org/[MongoDB] instance
+
+
+Fortunately, with Pipeline we have a couple of useful tools at our disposal:
+
+* Tool Auto-Installers
+* The link:https://go.cloudbees.com/docs/cloudbees-documentation/cje-user-guide/chapter-docker-workflow.html[CloudBees Docker Pipeline plugin]
+
+
+Tool Auto-Installers are exposed in Pipeline through the `tool` step and on
+link:https://ci.jenkins.io[ci.jenkins.io] we already had JDK8 and Groovy
+available:
+
+[source,groovy]
+----
+node('census && docker') {
+    /* .. */
+
+    def javaHome = tool(name: 'jdk8')
```

Comment:
One of the oft cited sleights against Ruby and Groovy is the ability to have this kind of ambiguous syntax IMO. For demonstration purposes I err on the side of strictness to make sure what is being done is (hopefully) as clear as possible